src/qililab/instruments/qdevil/qdevil_qdac2.py

```diff
@@ -102,6 +105,58 @@ def set_parameter(self, parameter: Parameter, value: ParameterValue, channel_id:
             return
         raise ParameterNotFound(self, parameter)
 
+    def get_dac(self, channel_id: ChannelID):
+        """Get specific DAC from QDAC.
+
+        Args:
+            channel_id (ChannelID): channel id of the dac
+        """
+        return self.device.channel(channel_id)
+
+    def upload_waveform(self, waveform: Waveform, channel_id: ChannelID):
+        """Uploads a waveform to the instrument and saves it to _cache.
+        IMPORTANT: note that the waveform resolution is not to the ns, it is acutally around 1_micro_second.
+
+        Args:
+            waveform (Waveform): Waveform to upload
+            channel_id (ChannelID): channel id of the qdac
+
+        Raises:
+            ValueError: if a waveform is already allocated
+        """
+        values = list(waveform.envelope())  # TODO: does np array work?
+        if channel_id in self._cache:
+            raise ValueError(
+                f"Device {self.name} already has a waveform allocated to channel {channel_id}. Clear the cache before allocating a new waveform"
+            )
+        trace = self.device.allocate_trace(channel_id, len(values))
```

Comment:
The QDAC manual says that uploading waveforms over LAN connection can sometimes cause errors due to 
unestability of the network, so they recommend to check after uploading if there are any errors, and if there are, re-upload the waveform again. I think we should check for that to make this as robust as possible.

"When uploading vast amount of binary data over LAN, for example multiple large (4-6 mio points)
arbitrary waveform traces), the instrument may occasionally report an error and the upload will fail. It
happens if there are some instabilities in the local network, but other times this is not the cause. So
always check for errors after uploading a trace, and in case of errors, re-upload the trace."